prow/federated_cluster_credentials.md

```diff
@@ -0,0 +1,224 @@
+| _service cluster_ | The _service cluster_ is the _context_ in which the `ProwJob` `CustomResourceDefinition` is deployed. |
+| _in-cluster context_ | The _in-cluster context_ is a _context_ for the `ServiceAccount` under which a `Pod` runs, loaded from a customary location in the `Pod`'s filesystem. |
+
+### Flags
+
+All Prow controllers that interact with resources expose the following flags:
+
+| Flag | Use |
+| ---- | --- |
+| `--kubeconfig` | Path to a `~/.kube/config` file to load. Can be passed more than once. |
+| `--use-explicit-credentials` | A feature flag to opt into the behavior in this document. |
+| `--prow-job-context` | The _context_ for the _service cluster_, used to interact with `ProwJob` `CustomResources`. |
+ 
+Note: until January 2020, the behavior in this document will be guarded behind the `--use-explicit-credentials` flag.
+
+## Configuring Credentials
+
+In order to configure federated credentials, Prow uses the `Config` format (often stored in `~/.kube/config`). Multiple `Config` files may be provided by passing `--kubeconfig` more than once. These files will be merged with a simple union on the _clusters_, _contexts_ and _users_ in each. If any names collide, loading will fail.
+
+## Defaulting
+
+When the `--kubeconfig` flag is not passed, Prow will use the _in-cluster context_ for all operations, like:
+
+ - creating and reading `ProwJob` `CustomResources`
+ - running `Pods` to execute `ProwJob`s that use the `kubernetes` agent
+ - executing Tekton pipelines for `ProwJob`s that use the `tekton` agent
+ 
+If a `--kubeconfig` flag is passed and the `--use-explicit-credentials` flag is set, no defaulting of any sort will occur. No _context_ is default and no _context_ is special in this mode. Jobs must opt into using a specific loaded _context_ by setting the `cluster` field. Prow components will continue using the _in-cluster context_ for interacting with `ProwJob` `CustomResources`.
+
+## Specifying an _in-cluster context_ via `--kubeconfig`
+
+Prow will post-process the `Config` loaded via `--kubeconfig` to allow users to specify the _in-cluster context_ but pass credentials by setting a `ServiceAccount` for the `Pod` in question. 
+
+In order to opt into this, use the `"in-cluster"` string in a _context_ and omit a _user_:
+
+```yaml
+contexts:
+- name: "my-in-cluster-context"
+  context:
+    cluster: "in-cluster"
+    namespace: "default"
+```
+
+## A Note On Namespaces
+
+Although the field used to target a _context_ for a `ProwJob` is named `cluster`, in reality the field targets a _cluster_ in a specific `Namespace` using a specific _user_. In keeping with Kubernetes' multi-tenant nature, Prow resources are namespaced and everything from the _service cluster_ to any given job may be sequestered in a unique `Namespace`. 
+
+An administrator may allow scheduling `Pod`s by `plank` in a specific `Namespace` by:
```

Comment:
Do we want to note that ensuring that jobs use the right contexts (aka job A uses context A and not context B) is done by validating the config files before uploading them into the cluster?

We need to make sure that the work to do in-repo config cannot violate this.

Reply:
Right now that's not a Prow concept, though, is it? It is a custom set of pre-validation logic for prow.k8s.io -- do we want to make a more general statement about how Prow should do this validation?

Reply:
Just pointing out that right now:
* We use presubmits to enforce that a job chooses an appropriate cluster context.
* The in-repo job definition -- including allowing jobs to be defined/redefined in presubmit -- breaks this assumption.

We should give people a best practice here: currently use a presubmit, not sure we've fully thought about this for in-repo job defs @alvaroaleman

Reply:
I am fairly certain the in-repo proposal includes a synthetic status that runs custom validation.

Reply:
Nice! Maybe link to the best current docs on how to write this custom validator?